src/outpack/location.py

```diff
@@ -71,6 +82,149 @@ def outpack_location_rename(old, new, root=None, *, locate=True):
     update_config(config, root.path)
 
 
+def location_resolve_valid(
+    location, root, include_local, include_orphan, allow_no_locations
+):
+    if location is None:
+        location = outpack_location_list(root)
+    elif isinstance(location, str):
+        if location not in outpack_location_list(root):
+            msg = f"Unknown location: '{location}'"
+            raise Exception(msg)
+        else:
+            location = [location]
+    elif isinstance(location, collections.abc.Iterable) and all(
+        isinstance(item, str) for item in location
+    ):
+        unknown = set(location).difference(outpack_location_list(root))
+        if len(unknown) > 0:
+            unknown_text = "', '".join(unknown)
+            msg = f"Unknown location: '{unknown_text}'"
+            raise Exception(msg)
+    else:
+        msg = (
+            "Invalid input for 'location'; expected None or a list of "
+            "strings"
+        )
+        raise Exception(msg)
+
+    if not include_local and LOCATION_LOCAL in location:
+        location.remove(LOCATION_LOCAL)
+    if not include_orphan and LOCATION_ORPHAN in location:  # pragma: no cover
+        location.remove(LOCATION_ORPHAN)
+
+    if len(location) == 0 and not allow_no_locations:
+        msg = "No suitable location found"
+        raise Exception(msg)
+
+    return location
+
+
+def outpack_location_pull_metadata(location=None, root=None, *, locate=True):
+    root = root_open(root, locate)
+    location_name = location_resolve_valid(
+        location,
+        root,
+        include_local=False,
+        include_orphan=False,
+        allow_no_locations=True,
+    )
+    for name in location_name:
+        location = OutpackLocation(name, root)
+        location.pull_all_metadata()
+        known_packets = []
+        for packet_location in root.index.data().location.values():
+            known_packets.extend(list(packet_location.values()))
+        location.validate_hashes(known_packets)
+        location.mark_known()
+
+    # TODO: deorphan recovered packets
+
+
+class OutpackLocation:
+    def __init__(self, location_name, root):
+        self.__root = root
+        self.location_name = location_name
+        self.__driver = _location_driver(location_name, root)
+        self._hint_remove = (
+            f'Probably all you can do at this point is '
+            f'remove this location from your configuration '
+            f'by running '
+            f'orderly_location_remove("{self.location_name}")'
+        )
+
+    def pull_all_metadata(self):
+        known_there = self.__driver.list()
+        known_here = self.__root.index.data().metadata.keys()
+        for packet_id in known_there:
+            if packet_id not in known_here:
+                self.pull_packet_metadata(packet_id)
+
+    def pull_packet_metadata(self, packet_id):
```

Comment:
should this one move into a private free function? is it part of the public API of a location?